### PR TITLE
Fixed missing arguments for bitstamp, kraken and coinmate exchanges in tradehistory api

### PR DESCRIPTION
**Pull request by generalbytes**
Fixed missing arguments for bitstamp, kraken and coinmate exchanges in tradehistory api

### CODE REVIEW

**Comment by timmolter** [comment]
I reverted the one dependency that was compiled with JDK 11 a while ago, so that issue should not be present anymore. What exact issue are you facing? Which dependency is the problem?

**Comment by generalbytes** [comment]
jitpack.yml file has been removed - not needed any more.
conflicts with current dev branch resolved.

**Comment by timmolter** [comment]
ERROR] /home/travis/build/knowm/XChange/xchange-bitstamp/src/main/java/org/knowm/xchange/bitstamp/service/BitstampAccountService.java:[181,83] incompatible types: java.lang.Long cannot be converted to java.lang.String

**Comment by generalbytes** [comment]
I noticed. I made mistake during merge to current develop branch. Working on fix now.

File: xchange-kraken/src/main/java/org/knowm/xchange/kraken/service/KrakenTradeService.java [hunk 1]
```diff
@@ -68,32 +68,44 @@ public boolean cancelOrder(CancelOrderParams orderParams) throws IOException {
...
+      }
+      if (idSpan.getEndId() != null) {
+        end = idSpan.getEndId();
+      }
+    }
+
+    if (params instanceof TradeHistoryParamsTimeSpan) {
+      TradeHistoryParamsTimeSpan timeSpan = (TradeHistoryParamsTimeSpan) params;
+      if (timeSpan.getStartTime() != null) {
+        start = DateUtils.toUnixTimeNullSafe(timeSpan.getStartTime()) + "";
```
Comment (walec51) [inline xchange-kraken/src/main/java/org/knowm/xchange/kraken/service/KrakenTradeService.java:100]: `+ ""` i unnecesary 

`getKrakenTradeHistory` should support null values

File: xchange-bitstamp/src/main/java/org/knowm/xchange/bitstamp/service/BitstampTradeServiceRaw.java [hunk 2]
```diff
@@ -20,20 +21,22 @@
   private final BitstampAuthenticated bitstampAuthenticated;
   private final BitstampAuthenticatedV2 bitstampAuthenticatedV2;
   private final BitstampDigest signatureCreator;
-  private final String apiKey;
-  private final SynchronizedValueFactory<Long> nonceFactory;
+  private String apiKey;
+  private SynchronizedValueFactory<Long> nonceFactory;
```
Comment (walec51) [inline xchange-bitstamp/src/main/java/org/knowm/xchange/bitstamp/service/BitstampTradeServiceRaw.java:25]: why is this not final any more?

File: xchange-kraken/src/main/java/org/knowm/xchange/kraken/service/KrakenTradeService.java [hunk 3]
```diff
@@ -68,32 +68,44 @@ public boolean cancelOrder(CancelOrderParams orderParams) throws IOException {
...
-    final Long offset;
     if (params instanceof TradeHistoryParamOffset) {
       offset = ((TradeHistoryParamOffset) params).getOffset();
-    } else {
-      offset = null;
+    }
+
+    if (params instanceof TradeHistoryParamsIdSpan) {
+      TradeHistoryParamsIdSpan idSpan = (TradeHistoryParamsIdSpan) params;
+      if (idSpan.getStartId() != null) {
```
Comment (walec51) [inline xchange-kraken/src/main/java/org/knowm/xchange/kraken/service/KrakenTradeService.java:89]: these `if`s are unnecessary

just doing `start = idSpan.getStartId();` for a null `idSpan.getStartId()` will have the same result in `start` as it is initialized to `null`

**Review by walec51 — CHANGES_REQUESTED** [review]

**Comment by rpanak-generalbytes** [comment]
@timmolter @walec51 I think all the mentioned problems should be fixed now.

**Comment by rpanak-generalbytes** [comment]
> I reverted the one dependency that was compiled with JDK 11 a while ago, so that issue should not be present anymore. What exact issue are you facing? Which dependency is the problem?

There is still some dependency issue, though. This is the JitPack log when building the PR (with merged upstream changes from today): https://jitpack.io/com/github/GENERALBYTESCOM/XChange/b88e82fab6/build.log

It seems like it's a maven plugin problem (fmt-maven-plugin). It depends on google-java-format lib whose published jar is built by JDK14.

**Review by walec51 — APPROVED** [review]

**Comment by milczarekIT** [comment]
Here is example json trade from Bitstamp:
`{"fee": "0.00000", "order_id": 1274846773324107, "datetime": "2020-09-14 09:47:45.577000", "usd": 0.0, "btc": "5.61134825", "btc_eur": 8751.00000000, "type": "2", "id": 123193752, "eur": "-49100.00"}`

Bitstamp uses ids as numbers (see `order_id` and `id` properties). 



There are also long types in `org.knowm.xchange.bitstamp.dto.trade.BitstampUserTransaction` (for `id` and `order_id`).

Why in this PR in `org.knowm.xchange.bitstamp.BitstampAuthenticatedV2` types has been changes from `Long` to `String` (@generalbytes , @walec51)?
I know that `org.knowm.xchange.dto.trade.UserTrade` and `org.knowm.xchange.dto.trade.Trade` have `orderId` and `id` with String type, but `org.knowm.xchange.bitstamp.BitstampAuthenticatedV2` is operating on `org.knowm.xchange.bitstamp.dto.trade.BitstampUserTransaction`.

**Comment by walec51** [comment]
@milczarekIT its returned as a string by the API and there is now guarantee in the docs what it will always by an integer

obraz

**Comment by milczarekIT** [comment]
@walec51, I was not writting about integers, but long types.

Presented output from postman is for public api (`/api/v2/transactions/btcusd`). Changes in this PR has been introduced for `/api/v2/user_transactions/`. There are no strings in Bitstamp json for `order_id` and `id`. I am pasting sample json from Bitstamp once again:
```
{"fee": "0.00000", "order_id": 1274846773324107, "datetime": "2020-09-14 09:47:45.577000", "usd": 0.0, "btc": "5.61134825", "btc_eur": 8751.00000000, "type": "2", "id": 123193752, "eur": "-49100.00"}
```

**Comment by walec51** [comment]
@milczarekIT long is just a short name for long integer

are you sure they won't change ids in /api/v2/user_transactions/ to match them up with those in /api/v2/transactions/... ?

they seam to be very loose on this matter, in the response you gave you have `"usd": 0.0` and `"eur": "-49100.00"`

interpreting id fields as string is safer in my opinion and more future proof when we have and exchange API that is very inconsistent in that matter, if one day they start returning UUID (like Bittrex has started to do) then deserialiaizing those ids as Longs will throw an exception

**Comment by generalbytes** [comment]
I suggest to leave order_id as a String. Bitstamp in past used numbers longer than java "long" type allows and it broke multiple implementations. It is better to stay on the safe side with String type.

**Comment by milczarekIT** [comment]
> long is just a short name for long integer

Yes, so integer and short integer can be used interchangeably, but long is a different type. I just wanted to clarify.

> are you sure they won't change

No, I can not (becasue I am not a creator of this API). 

Ok, I just wanted to clarify what was the purpose of change, let's leave id as string. @generalbytes, @walec51 are you going to change `long` ids for other exchanges (to be safe for API change)?

**Comment by walec51** [comment]
@milczarekIT 

when I talk if the id will be always an integer I meant it in the mathematical sends https://en.wikipedia.org/wiki/Integer - stop nit picking which Java type we should use for integer ids - yes `long` is the proper Java type for ids if you are certain that they will be always integer number - but that's not the point of the discussion

I don't think we have to use `String` for ids for all exchange specific DTOs - only for those who do not state in their documentation explicitly that their id is an integer number

**Pull request merged**